### DISCUSSION TITLE
What to do with SKR MINI

### DISCUSSION THREAD

**Discussion by SeanPluemer** [rdudhagra/Sand-Table#10] [Troubleshooting] [answered]
Good evening,

I have been able to complete everything up to actually getting the thing to move. 
I chose to use a SKR MINI rather than an Arduino. Kind of regretting this now, since I am unable to figure out why it won't move. 

I am able to get access to the GUI and upload a thr file, I can get all the way to playing it. It appears that it is playing, but there is no movement. 
I will be trying to learn how this whole code works tomorrow, (thank you, by the way, this is amazing that you are sharing this) and see if I can figure anything out. But in the meantime, is there anything that needs to be changed for use of the SKR MINI v2?

**Comment by rdudhagra** [comment]
Hi Sean, a few questions to troubleshoot...I assume you’re still using the TMC2209 drivers. If not, there will be more things that you’ll need to change:
1. Did you configure the board by placing the jumpers etc per the YouTube video I linked to in the build guide (https://youtu.be/dOJbSrWVu_Q?t=439)? The TMC2209 drivers are special and will only work if they’re properly wired.
2. You’ll have to modify the marlin arduino code’s configuration to use the SKR MINI V2 instead of the RAMPS board, not sure if you’ve done that yet. I can send pointers on how to do that if you can’t figure it out.
3. Download a software like pronterface onto your computer and try to control the board. Do the motors move at all? Does your computer even connect to the board?
4. If step 3 works, your serial ports on the Pi might be switched/incorrect. Check that the serial port numbers defined in `.env` do indeed match. You can do this by unplugging everything and running `ls /dev/tty*`, then plugging in the SKR board and running that command again. There should be one new entry, which is the serial port of the SKR. You can do the same for the LED Strip Arduino as well. (I think I forgot to mention this in the build guide...I’ll put this blurb in there as well to avoid future confusion).

Let me know if those steps help. If not, we might have a more nuanced problem on our hands...keep me posted!

**Reply by SeanPluemer** [reply]
Thank you for the quick responses!

1. Yes, I believe the jumpers are configured correctly. 
2. I have not looked into doing this. Pointers would be amazing, and in case anyone else has this issue they could reference them too.
3. I was able to download pronterface and was able to control the board. I was able to move the motors, and even load in a gcode! Very exciting to see everything working. 
4. I did not have enough time earlier to test this (got too distracted playing with pronterface). I will check this evening and update you.


Thank you so much!
Sean

**Reply by rdudhagra** [reply]
That's great news! If everything seems to move, then I don't think there is a need to update the firmware (Although for future reference here's a link on changing the motherboard in Marlin: https://marlinfw.org/docs/configuration/configuration.html#motherboard). After all, we're only using 5% of Marlin's capability (no extruders, homing switches, LCD displays, SD card slot, etc...), so I don't think there will be any issues. I suspect it's step 4 that is the culprit, though. Let me know if that ends up being it, and I'll update the build guide readme with the above instructions.

**Reply by SeanPluemer** [reply]
Greetings, 
I was able to test step 4. I tried a few different ways, but the reoccurring theme is, if I start the pi with the SKR plugged in, it shows up as ACM0. Then if I unplug and plug back in the SKR it will change to ACM1.
What is just swap the two in the .env file. So now the Marlin = ACM0 and the LED = ACM1.

So, when I plug in the Pi and SKR the carriage will move a foot to the right and that's it. After that, nothing I do will get it to move again. I have tried 3 diff .thr files with no luck. 
This is kind of an off-question, I do not have an Arduino powering LED's rn, could this be an issue if I'm not using an Arduino?

**Reply by rdudhagra** [reply]
Hmm, a few more questions:

1. Are you plugging the SKR into the same physical USB port on the Pi every time? I think the ACM number depends on which port the device is plugged into.
2. When you control the SKR using pronterface, are you able to run the homing sequence? The first thing the pi does is trigger the skr to home itself, and it could be that it’s getting stuck during that process. If this is the case, you’ll need to modify the sensitivity of the homing numbers in Marlin, which the above linked YouTube video should go through on how to do that. Also, just to confirm again, you should have wires that go from one of the top pins on the tmc2209’s to somewhere on the SKR. That’ll enable sensorless homing.
3. If the above works, rerun the sand table and after the carriage freezes again, paste the output of `sudo systemctl status sandtable` and `journalctl -xe`. Could be that the website is freaking out for some odd reason.

**Reply by SeanPluemer** [reply]
1. Yes I am plugging it into the same port.
2. I think this might be what the issue is. When I try to home using pronterface it will move for a bit and I get an "[ERROR] Error:Printer halted. kill() called!" I have reviewed the above video and I believe it is set up correctly, but this is a slight issue because I ordered [this](https://www.amazon.com/gp/product/B089CQ3DHB/ref=ppx_yo_dt_b_asin_title_o05_s02?ie=UTF8&psc=1) SKR V2 with the integrated tmc2209's. I believe I saw documentation somewhere that said that no modification was needed except adding the 4 jumpers on the top. 
3. I have attached the outputs below.

Systemctl status: 
![image](https://user-images.githubusercontent.com/58951141/108640344-565ee080-7456-11eb-840c-e2a9aace2e90.png)
journalctl:
![image](https://user-images.githubusercontent.com/58951141/108640363-70002800-7456-11eb-9a57-385561c1e1a9.png)

**Reply by SeanPluemer** [reply]
I found [this document ](https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/blob/master/hardware/BTT%20SKR%20MINI%20E3%20V2.0/Hardware/BTT%20SKR%20MINI%20E3%20V2.0%20Instruction%20Manual.pdf) that shows how the sensorless homing should be setup for the SKR v2.
![image](https://user-images.githubusercontent.com/58951141/108640702-768f9f00-7458-11eb-983b-25196b8d791d.png)

**Reply by rdudhagra** [reply]
Thanks for the updates...yea this document helps a lot. I assume you have those jumpers in place, at least for the X and Y drivers?

**Reply by rdudhagra** [reply]
The issue is definitely with the homing then, since you got an error when trying to send `G28`. Have you made any modifications to the marlin firmware I have on my repo, and if so what have you changed? (want to make sure we're on the same page)

If you haven't done already, can you modify my sand-table Marlin firmware in the following ways, upload that to the SKR, and retry `G28`? (line numbers might be slightly off since I'm pulling these figures from the [SKR's official marlin firmware build](https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/blob/master/firmware/V2.0/Marlin-2.0.7.2-SKR-mini-E3-V2.0/Marlin/Configuration.h)
1. Configuration.h (line 132) `#define MOTHERBOARD BOARD_BTT_SKR_MINI_E3_V2_0`
2. Configuration_adv.h (line 2516) `#define SENSORLESS_HOMING // StallGuard capable drivers only
` (should be uncommented, just to confirm)

If this doesn't work, try installing whatever official build of Marlin exists for the SKR (I assume the one from bigtreetech that I linked above). If that ends up working, then all you'll have to do is change the size of your print area and maybe enable more micro stepping, but I can walk you through what to change if you get to that point.

**Reply by SeanPluemer** [reply]
Hello,
Still no luck, but here is where I am at. I was unable to get your Marlin to work. I am now using [this](https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/tree/master/firmware/V2.0) Marlin. I was able to configure it and get everything working using pronterface. I can home, and run GCODE now using pronterface. 
Now my issue I can't get the raspberry pi to communicate with it. When I boot up the pi, it does appear to try to home but I  don't know if it passes or not. 
Feel free to email me at  pluemersean@gmail.com if you want to have a bit quicker communication    . 
I have forked this my [repository](https://github.com/SeanPluemer/Sand-Table), I believe this is set up to allow you to view it. Forgive me, a noob with GitHub if not.

**Comment by SeanPluemer** [comment, marked as answer]
This issue was resolved over the phone. What we did to resolve the issue was do the following:

1. Install the correct Marlin software from [here](https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3).
2. Enable and tune the sensorless homing. 
3. Enable CoreXY mode on Marlin.
4. Modify the home command (when the pi first boots up) to only home the X and Y axis. (for some reason the SKR was getting hung up trying to home a non-existing Z-axis.

**Reply by SeanTurner71** [reply]
I'm just starting on installing Marlin on a skr mini V2.  Was wondering if you could pass along the info that you gave him?  Thanks in advance.

**Reply by SeanPluemer** [reply]
Hello Sean (nice name btw),

So you are going to get to learn to program the SKR! 
I **think** my fork has all the changes you will need.... You can view my Marlin [here](https://github.com/SeanPluemer/Sand-Table/tree/master/Marlin-2.0.x/Marlin). So in theory, all you need to do is to download my Marlin and flash it to the board... But it has been a few months and I don't remember tbh. 
Here is a video that shows how to flash the board and what to change to enable sensorless homing. https://www.youtube.com/watch?v=0aizXlcQGPA

You can use a software on your laptop called pronterface to make sure that everything works correctly before you even try dealing with the Pi. 

The last thing that you will need to change is the raspberry pi setup g28 code.
This can be found at Sand-Table/Website/server/routes/gcode_controller.js linen ~352. 
There will be a line that just says "G28"  I think all you need to do is change it to "G28 X Y" and that will make ur life

I think that was all I had to do to get it working (other than changing the size of my table). But if not, feel free to msg me!

**Comment by SeanTurner71** [comment]
Wow.  Thanks alot.  I'm not a total noob on Marlin but it's been a while
and I have moved onto Klipper <https://www.youtube.com/watch?v=3-_EHVw4-qw>.
I couldn't get your Marlin to compile for some reason so I just looked
through your files and changed the settings into the newer version.  I got
to see movement using pronterface (which is great) but it's not using
corexy for some reason, I defined it in marlin but it's only running one
motor at a time.  I have some troubleshooting to do I guess.  Thanks again
for giving me a heads up on the G28 command editing also.

Sean Turner

**Reply by SeanPluemer** [reply]
So since it's not using corexy that almost makes me think that the firmware didn't flash.
In pronterface, you can home the axis. Try to home the x and then the y axis don't hit the home button on the bottom left or it will lock out the skr. Just hit the home on the x (top left) and y(top right) separately. The goal for this is to see if the sensorless homing is actually working. If it gets to the end and starts grinding away, sounds like the updated firmware didn't flash.

**Reply by SeanTurner71** [reply]
I've tried so many things at the moment I don't know which way is up.  First thing first.  From what I have read the V2 board does not need any special wiring to enable sensorless homing right?  Just install the jumper right?  Since its moving diagonally all the time I'm getting mixed signals on the homing.

EDIT:  Ok if I hit the red center button in pronterface it will move in core xy mode and if I send a g28 it will move in straight lines.  I guess that is something

**Reply by SeanTurner71** [reply]
EDIT 2: sensorless homeing does not seem to work.  $50 to whoever can fix this.  The sounds of belts jumping is going to haunt my dreams

**Reply by SeanTurner71** [reply]
Final Edit:.  Bad flash card that came with the board.  It works now.  I'm gonna go drink now.  Thanks for help SeanPluemer

**Reply by SeanPluemer** [reply]
Haha, I can relate to that! Glad u got it working!

**Reply by rdudhagra** [reply]
Thanks @SeanPluemer  for helping out on this one! So I guess long-story-short don't use garbage SD cards :P

**Reply by SeanTurner71** [reply]
Since this post I found out what was actually causing my issue so I thought I would update it in case it happened to someone else.  It turned out that by leaving the SKR plugged into my PC while powering it on was causing it to fail to boot properly.  I keep it plugged in to use pronterface and that was my mistake.  When I changed flash cards I just happened to have it unplugged.
Hopefully that will be of help to someone else.

**Reply by gabhalling** [reply]
Hey guys, I know it's been a while since you worked on this project but I'm attempting to make it now and I'm having some issues you seem to have gone through.

For some reason, VSCode does not seem to be able to compile the Marlin Directory from SeanPluemer (I'm also using the BTT SKR Mini E3 V2. I get this message: 

**********
RuntimeError: deque mutated during iteration:
  File "C:\Users\gabha\.platformio\penv\Lib\site-packages\platformio\builder\main.py", line 180:
    env.SConscript(env.GetExtraScripts("post"), exports="env")
  File "C:\Users\gabha\.platformio\packages\tool-scons\scons-local-4.7.0\SCons\Script\SConscript.py", line 612:
    return _SConscript(self.fs, *files, **subst_kw)
  File "C:\Users\gabha\.platformio\packages\tool-scons\scons-local-4.7.0\SCons\Script\SConscript.py", line 279:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "C:\Users\gabha\Downloads\Sand-Table-master\Marlin-2.0.x\buildroot\share\PlatformIO\scripts\STM32F103RC_SKR_MINI.py", line 6:
    for define in env['CPPDEFINES']:
**********

SeanTurner71 mentionned something about changing the settings to the newer version, but I am unsure on the way to do it.

Any help will be greatly appreciated!

**Reply by gabhalling** [reply]
In case someone else also ends up in this situation, I ended up starting with the BTT Marlin Build for the SKR Mini E3 V2 (https://github.com/bigtreetech/BIGTREETECH-SKR-mini-E3/tree/master/firmware/V2.0/Marlin-2.0.8.2.x-SKR-mini-E3-V2.0) and modifying values such as serial ports, bed size, enabling CoreXY, sensorless homing and tuning it.

But I was still getting an error (RuntimeError: deque mutated during iteration:) when I tried to compile and I had to follow eoyilmaz's advice that I found here: https://community.platformio.org/t/marlin-runtimeerror-deque-mutated-during-iteration/34661

On to the next step now!

**Reply by rdudhagra** [reply]
Just checking in here...are you still running into issues with Marlin?

**Reply by gabhalling** [reply]
As of now, not anymore. Might have to do some tuning for the sensorless homing when I get to it but besides that, I got it to build and flashed the board so that I was able to move the motors.
